### PR TITLE
Feature/set email as ref ID

### PR DESCRIPTION
**Pull request by jamesmengo**
## Overview ⚙️

Setting the Firebase document ref ID to the user's email. After more research I feel like we should actually go in a different direction, since I have a better understanding of firebase auth now. I feel like we should be using the User UID provided by firebase auth instead of the emails. Refer to the discussion section for more info.

This change
- Sets the user firebaseID to their email
- In each event, the ID of owners and users will be their email

##Discussion##
- Google Auth provides a User UID when you authenticate - should we be using this instead? We can explicitly set our document reference ID to the provided UID so I think this will address the problem that  we were talking about @gokcedilek  (Where the ID for schema is different from user). This makes more sense to me, but I don't have that much experience doing this stuff.

Thoughts?

image

Closes #1234

### CODE REVIEW

**Comment by nandanmen** [comment]
I'm thinking one drawback to this is that users will be able to create a separate account using their gmail if they don't choose "login via Google". Most of the apps I use redirect users to their Google account if they've previously logged in with google and try to make an account with the same email.

This is a bit out of scope for MVP, but I think something we should think about.

**Comment by jamesmengo** [comment]
> I'm thinking one drawback to this is that users will be able to create a separate account using their gmail if they don't choose "login via Google". Most of the apps I use redirect users to their Google account if they've previously logged in with google and try to make an account with the same email.

From my understanding firebase auth already behaves this way. I tested this by trying to create an email/password account that already had been registered via google auth.

The flow I'm thinking of:
1) Log in / create an account to access your User UID - both google auth and email/password
2) Use your User UID to query your actual user data that's stored, as well as events etc

image




Maybe we can consider using tokens. Haven't had enough time to give it proper thought unfortunately :/

**Comment by jamesmengo** [comment]
@narendrasss  let me do some more reading later tonight and get back to you

**Comment by jamesmengo** [comment]
https://firebase.google.com/docs/auth/admin/verify-id-tokens

**Comment by gokcedilek** [comment]
Yes, it also makes more sense to me at least for now to use login uid as the doc id. It seems like if we first create a gmail account, and then try to signin with google, firebase does not override the uid, so we can rely on the uid. If we first signin with google, and then try to create a gmail account, we get an "auth/email-already-in-use" error, which I guess is the desired behavior so the existing email account is not overridden/duplicated (I'm not sure how exactly we would do that, but for example if we catch this error with a gmail account, we can look into redirecting to google sign in)

**Comment by jamesmengo** [comment]
@gokcedilek I think we should do this in 2 phases. For now I would prefer to use email since it will be easier for us to test and develop. Once front-end can send tokens we can convert our endpoints to use those instead maybe a WT2 thing.

File: backend/data/schema/schema.go [hunk 1]
```diff
@@ -11,6 +11,7 @@ type User struct {
 	FirebaseID string
 	Username   string
 	Email      string
+	Events		[]string // FirebaseID of events that this user is attending or the owner of
```
Comment (jamesmengo) [inline backend/data/schema/schema.go:14]: This is just a tiny change I snuck in here. I can make a seperate PR just lmk.

Comment (jamesmengo) [reply backend/data/schema/schema.go:14]: nvm github is struggling and the diff is incorrect - this change is already merge in. The change in this file is a removed comment that I added

File: backend/data/clients/firebase/write.go [hunk 2]
```diff
@@ -2,17 +2,24 @@ package firebase
...
 )
 
 func (u *User) Write() error {
 	// todo: consider logging ref or actual document info here
 	log.Infof("writing user to firestore: %+v", u)
-
-	ref, _, err := Client.Collection("users").Add(context.Background(), u)
-	u.FirebaseID = ref.ID
-
+	// check if user with provided email already exists
```
Comment (jamesmengo) [inline backend/data/clients/firebase/write.go:13]: For now, for ease of development I say that we just use Email. Setting it to the UID will require tokens to be sent. We can implement that in phase 2 after we have basic functionality working better

**Review by jackyzha0 — APPROVED** [review]

**Review by nandanmen — APPROVED** [review]

File: backend/data/schema/schema.go (hunk 1 above)
Comment (nandanmen) [reply backend/data/schema/schema.go:14]: Might have to merge in the latest master for this diff to show up correctly - not a huge deal though.

**Pull request merged**